devito/finite_differences/differentiable.py

```diff
@@ -797,9 +797,10 @@ def _evaluate(self, **kwargs):
         f = w.function
         d = w.dimension
         mapper = {w.subs(d, i): f.weights[n] for n, i in enumerate(d.range)}
+        basexpr = set(a.function for a in expr._functions if a.is_Function) - {f}
```

Comment:
What happens if you have `(u*v).dx` ? why do you only `pop` one (below)?

We could also perform a stronger error check here...